Defining a closure in a macro causes a compiler exception

## Compiler version

3.0.0-RC3

## Minimized example

```scala
import scala.quoted.*

object TestMacro {
  def use(f: () => String): Unit = ()

  inline def test: Unit = ${testImpl}

  def testImpl(using Quotes): Expr[Unit] = {
    import quotes.reflect.*

    def resultDefBody(): Term = '{
      val result: String = "xxx"
      result
    }.asTerm
    val resultDefSymbol = Symbol.newMethod(Symbol.spliceOwner, "getResult", MethodType(Nil)(_ => Nil, _ => TypeRepr.of[String]))
    val resultDef = DefDef(resultDefSymbol, { case _ => Some(resultDefBody()) })
    val resultExpr = Block(List(resultDef), Closure(Ref(resultDefSymbol), None)).asExprOf[() => String]

    //

    val r = '{ TestMacro.use($resultExpr) }
    println(r.asTerm.show(using Printer.TreeShortCode))
    r
  }
}
```

and then calling the macro:

```scala
object Test extends App {
  TestMacro.test
}
```

## Output

I'm trying to generate code which passes a function as a parameter to a method. From what I've seen by printing the AST of quoted code, I need a `Block` with a `DefDef` as the statement, and a `Closure` as the expression. In the body of the function, I define a local variable - and this seems to be a problem. The original exception from the compiler I get is:

```
java.util.NoSuchElementException: val result while compiling Test.scala
[error] ## Exception when compiling 25 sources to /core/target/jvm-3.0.0-RC3/test-classes
[error] java.util.NoSuchElementException: val result
[error] scala.collection.mutable.AnyRefMap$ExceptionDefault.apply(AnyRefMap.scala:508)
[error] scala.collection.mutable.AnyRefMap$ExceptionDefault.apply(AnyRefMap.scala:507)
[error] scala.collection.mutable.AnyRefMap.apply(AnyRefMap.scala:207)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder$locals$.load(BCodeSkelBuilder.scala:507)
```

while trying to minify the example, I started getting another exception, so I'm not sure if my code reproduces the problem, but maybe it reproduces some problem :)

Here's the exception I get now:

```
java.lang.IllegalArgumentException: Could not find proxy for val result: String in List(val result, val <local Test$>, module class Test$, module class softwaremill, module class com, module class <root>), encl = (...)
[error] ## Exception when compiling 6 sources to /target/jvm-3.0.0-RC3/classes
[error] java.lang.IllegalArgumentException: Could not find proxy for val result: String in List(val result, val <local Test$>, module class Test$, module class softwaremill, module class com, module class <root>), encl = (...)
[error] dotty.tools.dotc.transform.LambdaLift$Lifter.searchIn$1(LambdaLift.scala:396)
[error] dotty.tools.dotc.transform.LambdaLift$Lifter.proxy(LambdaLift.scala:409)
[error] dotty.tools.dotc.transform.LambdaLift$Lifter.proxyRef(LambdaLift.scala:427)
[error] dotty.tools.dotc.transform.LambdaLift$Lifter.addFreeArgs$$anonfun$1(LambdaLift.scala:433)
[error] scala.collection.immutable.List.map(List.scala:246)
```

When printing the generated code, it looks fine:

```scala
TestMacro.use((() => {
  val result: String = "xxx"

  (result: String)
}))
```

## Expectation

A way to generate a closure and pass it to a method, in a macro.

Compiling this code using the `-Xcheck-macros` produces the following error

```scala
2 |  TestMacro.test
  |  ^^^^^^^^^^^^^^
  |Exception occurred while executing macro expansion.
  |java.lang.AssertionError: assertion failed: Tree had an unexpected owner for val result
  |Expected: method getResult (Test$._$_$getResult)
  |But was: val macro (Test$._$macro)
  |
  |
  |The code of the definition of val result is
  |val result: scala.Predef.String = "xxx"
  |
  |which was found in the code
  |{
  |  val result: scala.Predef.String = "xxx"
  |
  |  (result: java.lang.String)
  |}
  |
  |which has the AST representation
  |Inlined(Some(TypeIdent("TestMacro$")), Nil, Block(List(ValDef("result", TypeIdent("String"), Some(Literal(StringConstant("xxx"))))), Typed(Ident("result"), Inferred())))
  |
  |
  |     at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
  |     at scala.quoted.runtime.impl.QuotesImpl$$anon$9.traverse(QuotesImpl.scala:2825)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1584)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1584)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.fold$1(Trees.scala:1459)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.apply(Trees.scala:1461)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1492)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1585)
  |     at scala.quoted.runtime.impl.QuotesImpl$$anon$9.traverse(QuotesImpl.scala:2828)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1584)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1584)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1512)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1465)
  |     at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1585)
  |     at scala.quoted.runtime.impl.QuotesImpl$$anon$9.traverse(QuotesImpl.scala:2828)
  |     at scala.quoted.runtime.impl.QuotesImpl$reflect$.xCheckMacroOwners(QuotesImpl.scala:2829)
  |     at scala.quoted.runtime.impl.QuotesImpl$reflect$.scala$quoted$runtime$impl$QuotesImpl$reflect$$$xCheckMacroedOwners(QuotesImpl.scala:2791)
  |     at scala.quoted.runtime.impl.QuotesImpl$reflect$DefDef$.apply$$anonfun$2$$anonfun$1(QuotesImpl.scala:255)
  |     at dotty.tools.dotc.ast.tpd$.DefDef(tpd.scala:283)
  |     at scala.quoted.runtime.impl.QuotesImpl$reflect$DefDef$.apply$$anonfun$1(QuotesImpl.scala:256)
  |     at scala.quoted.runtime.impl.QuotesImpl$reflect$.scala$quoted$runtime$impl$QuotesImpl$reflect$$$withDefaultPos(QuotesImpl.scala:2782)
  |     at scala.quoted.runtime.impl.QuotesImpl$reflect$DefDef$.apply(QuotesImpl.scala:256)
  |     at scala.quoted.runtime.impl.QuotesImpl$reflect$DefDef$.apply(QuotesImpl.scala:253)
  |     at TestMacro$.testImpl(Macro_1.scala:16)
  |
  | This location contains code that was inlined from Test_2.scala:2
```

One simple way to fix the code is to change the owner of the content of `getResult`

```diff
-    val resultDef = DefDef(resultDefSymbol, { case _ => Some(resultDefBody()) })
+    val resultDef = DefDef(resultDefSymbol, { case _ => Some(resultDefBody().changeOwner(resultDefSymbol)) })
```

--- a/library/src/scala/quoted/Quotes.scala
+++ b/library/src/scala/quoted/Quotes.scala
@@ -497,6 +497,14 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
 
     /** Methods of the module object `val DefDef` */
     trait DefDefModule { this: DefDef.type =>
+      /** Create a method definition `def f[..](...)` with the signature defined in the symbol.
+       *
+       *  The `rhsFn` is a function that receives references to its parameters and should return
+       *  `Some` containing the implementation of the method. Returns `None` the method has no implementation.
+       *  Any definition directly inside the implementation should have `symbol` as owner.
+       *
+       *  See also: `Tree.changeOwner`
+       */
       def apply(symbol: Symbol, rhsFn: List[List[Tree]] => Option[Term]): DefDef
       def copy(original: Tree)(name: String, paramss: List[ParamClause], tpt: TypeTree, rhs: Option[Term]): DefDef
       def unapply(ddef: DefDef): (String, List[ParamClause], TypeTree, Option[Term])
@@ -558,6 +566,14 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
 
     /** Methods of the module object `val ValDef` */
     trait ValDefModule { this: ValDef.type =>
+      /** Create a value definition `val x`, `var x` or `lazy val x` with the signature defined in the symbol.
+       *
+       *  The `rhs` should return be `Some` containing the implementation of the method.
+       *  Returns `None` the method has no implementation.
+       *  Any definition directly inside the implementation should have `symbol` as owner.
+       *
+       *  See also: `Tree.changeOwner`
+       */
       def apply(symbol: Symbol, rhs: Option[Term]): ValDef
       def copy(original: Tree)(name: String, tpt: TypeTree, rhs: Option[Term]): ValDef
       def unapply(vdef: ValDef): (String, TypeTree, Option[Term])

--- a/tests/pos-macros/i12309/Macro_1.scala
+++ b/tests/pos-macros/i12309/Macro_1.scala
@@ -1,0 +1,25 @@
+import scala.quoted.*
+
+object TestMacro {
+  def use(f: () => String): Unit = ()
+
+  inline def test: Unit = ${testImpl}
+
+  def testImpl(using Quotes): Expr[Unit] = {
+    import quotes.reflect.*
+
+    def resultDefBody(): Term = '{
+      val result: String = "xxx"
+      result
+    }.asTerm
+    val resultDefSymbol = Symbol.newMethod(Symbol.spliceOwner, "getResult", MethodType(Nil)(_ => Nil, _ => TypeRepr.of[String]))
+    val resultDef = DefDef(resultDefSymbol, { case _ => Some(resultDefBody().changeOwner(resultDefSymbol)) })
+    val resultExpr = Block(List(resultDef), Closure(Ref(resultDefSymbol), None)).asExprOf[() => String]
+
+    //
+
+    val r = '{ TestMacro.use($resultExpr) }
+    // println(r.asTerm.show(using Printer.TreeShortCode))
+    r
+  }
+}

--- a/tests/pos-macros/i12309/Test_2.scala
+++ b/tests/pos-macros/i12309/Test_2.scala
@@ -1,0 +1,3 @@
+object Test extends App {
+  TestMacro.test
+}

